File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        arr = []
        n = len(asteroids)

        def remove(i, arr):
            while len(arr) > 0:
                N = len(arr)
                last = arr[N-1]
                
                if last < 0:
                    break
                
                if last == i:
                    arr.pop()
                    return
                
                if last > i:
                    return

                if last < i:
                    arr.pop()
            
            arr.append(-1 * i)

        for i in range(n):
            val = asteroids[i]
            if val < 0:
                remove(abs(val), arr)
            else:
                arr.append(val)
        return arr

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    void remove(int i, vector<int>& arr) {
        while (!arr.empty()) {
            int last = arr.back();

            if (last < 0) {
                break;
            }

            if (last == i) {
                arr.pop_back();
                return;
            }

            if (last > i) {
                return;
            }

            if (last < i) {
                arr.pop_back();
            }
        }
        arr.push_back(-1 * i);
    }

    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> arr;
        int n = asteroids.size();

        for (int i = 0; i < n; i++) {
            if (asteroids[i] < 0) {
                remove(abs(asteroids[i]), arr);
            } else {
                arr.push_back(asteroids[i]);
            }
        }
        return arr;
    }
};